mm2src/mm2_main/src/lp_ordermatch/orderbook_rpc.rs

```diff
@@ -140,6 +144,7 @@ pub async fn orderbook_rpc(ctx: MmArc, req: Json) -> Result<Response<Vec<u8>>, S
 
     try_s!(subscribe_to_orderbook_topic(&ctx, &base_ticker, &rel_ticker, request_orderbook).await);
     let orderbook = ordermatch_ctx.orderbook.lock();
+    let my_orders_pubkeys = ordermatch_ctx.my_p2p_pubkeys.lock();
```

Comment:
We can clone the `my_p2p_pubkeys` collection to avoid deadlocks

Reply:
As an alternative, we can try moving `my_p2p_pubkeys` to the `orderbook`. It will fit the `Orderbook` structure context.

Reply:
Cloning MutexGuard should be safe.

If we can do something  like:
```diff
diff --git a/mm2src/mm2_main/src/lp_ordermatch/orderbook_rpc.rs b/mm2src/mm2_main/src/lp_ordermatch/orderbook_rpc.rs
index dbebf4078..484cc4f9f 100644
--- a/mm2src/mm2_main/src/lp_ordermatch/orderbook_rpc.rs
+++ b/mm2src/mm2_main/src/lp_ordermatch/orderbook_rpc.rs
@@ -311,7 +311,6 @@ pub async fn orderbook_rpc_v2(
         .map_to_mm(OrderbookRpcError::P2PSubscribeError)?;

     let orderbook = ordermatch_ctx.orderbook.lock();
-    let my_orders_pubkeys = ordermatch_ctx.my_p2p_pubkeys.lock();
     let my_pubsecp = ctx.secp256k1_key_pair_as_option().map(|_| {
         CryptoCtx::from_ctx(&ctx)
             .expect("ctx is available")
@@ -337,7 +336,7 @@ pub async fn orderbook_rpc_v2(
                         continue;
                     },
                 };
-                let is_mine = is_my_order(&my_orders_pubkeys, &my_pubsecp, &ask.pubkey);
+                let is_mine = is_my_order(&ordermatch_ctx.my_p2p_pubkeys.lock(), &my_pubsecp, &ask.pubkey);
                 orderbook_entries.push(ask.as_rpc_v2_entry_ask(address, is_mine));
             }
             orderbook_entries
@@ -367,7 +366,7 @@ pub async fn orderbook_rpc_v2(
                         continue;
                     },
                 };
-                let is_mine = is_my_order(&my_orders_pubkeys, &my_pubsecp, &bid.pubkey);
+                let is_mine = is_my_order(&ordermatch_ctx.my_p2p_pubkeys.lock(), &my_pubsecp, &bid.pubkey);
                 orderbook_entries.push(bid.as_rpc_v2_entry_bid(address, is_mine));
             }
             orderbook_entries

```

We can avoid leaving `my_p2p_pubkeys` as locked for longer than we need.

Reply:
Since we lock the `OrdermatchContext::orderbook` mutex, we can be sure that it won't change during this function executes. Hence the `my_p2p_pubkeys` collection won't change.
Am I missing something?
> Cloning can introduce data inconsistency.

Reply:
@sergeyboyko0791 These mutexes are not locked simultaneously in all cases. For example:
https://github.com/KomodoPlatform/atomicDEX-API/blob/33fdfe8ff64093744454587da5babd9f07019ed1/mm2src/mm2_main/src/lp_ordermatch.rs#L4479-L4490

So here, `orderbook` is locked in `maker_order_created_p2p_notify` and a new order is added there. It seems that we already have data inconsistency because there is a time window when order is already in the `Orderbook`, but it's pubkey is not in `my_p2p_pubkeys`, so we can get a wrong `is_mine:false` from RPC.

I think it's one more reason to move `my_p2p_pubkeys` to `Orderbook`, create fn like
```rust
fn insert_or_update_order_my(ctx: &MmArc, item: OrderbookItem, my_order: &MakerOrder) {
    let ordermatch_ctx = OrdermatchContext::from_ctx(ctx).expect("from_ctx failed");
    let mut orderbook = ordermatch_ctx.orderbook.lock();
    orderbook.insert_or_update_order_update_trie(item);
    if let Some(key) = my_order.p2p_privkey {
        orderbook.my_p2p_pubkeys.insert(hex::encode(key.public_slice()));
    }
}
```

And use it in `maker_order_created_p2p_notify`.

Reply:
Now I see what you meant by `data inconsistency`!

I like this idea
> I think it's one more reason to move my_p2p_pubkeys to Orderbook, create fn like

Reply:
@laruh could you please implement the forementioned change suggestion? 🙂

Reply:
Moved my_p2p_pubkeys to Orderbook.
In accordance with the addition in `maker_order_created_p2p_notify`, they are removed in `maker_order_cancelled_p2p_notify`. In addition, I decided to keep adding pubkeys in [`orders_kick_start`](https://github.com/KomodoPlatform/atomicDEX-API/commit/6ede11586210d5fe641f0f1fb16844586b7a0c0f#diff-4c4940ed0030a5a26cc593870df06f290339e4ed69d0e7564b2c71135e4d2787R5157) function.